relentless/core/variable.py

```diff
@@ -0,0 +1,395 @@
+
+        Returns
+        -------
+        bool
+            True if the variable is at the lower bound, False otherwise.
+
+        """
+        return self.state is DesignVariable.State.LOW
+
+    def athigh(self):
+        """Confirms if the variable is at the upper bound.
+
+        Returns
+        -------
+        bool
+            True if the variable is at the upper bound, False otherwise.
+
+        """
+        return self.state is DesignVariable.State.HIGH
+
+class DependentVariable(Variable):
+    """Abstract base class for a variable that depends on other variables.
+
+    A dependent variable is composed from other variables. In addition to the
+    value property, it must also supply a derivative method with respect to
+    any of its dependent variables.
+
+    The number of dependencies of the variable are locked at construction, but
+    their values can be modified.
+
+    Parameters
+    ----------
+    vardicts : dict
+        Attributes for the variable on which the DependentVariable depends (positional argument).
+    kwvars : kwargs
+        Attributes for the variable on which the DependentVariable depends (keyword arguments).
+
+    """
+    def __init__(self, *vardicts, **kwvars):
+        attrs = {}
+        for d in vardicts:
+            attrs.update(d)
+        attrs.update(**kwvars)
+
+        if len(attrs) == 0:
+            raise AttributeError('No attributes specified for DependentVariable.')
+
+        for k,v in attrs.items():
+            super().__setattr__(k,self._assert_variable(v))
+        self._params = tuple(attrs.keys())
+
+    def __setattr__(self, name, value):
+        #Sets the value of the variable on which the specified DependentVariable depends.
+        if name != '_params' and name in self._params:
+            value = self._assert_variable(value)
+        super().__setattr__(name,value)
+
+    @property
+    def depends(self):
+        """Generator of parameter,object pairs."""
+        for p in self._params:
+            yield p,getattr(self,p)
+
+    @property
+    def params(self):
+        return self._params
+
+    def dependency_graph(self):
+        # construct graph of variables
+
+        # discover all variables (nodes) by quasi-depth first search
+        g = nx.DiGraph()
+        stack = [self]
+        while stack:
+            a = stack.pop()
+            if a not in g:
+                g.add_node(a)
+                if isinstance(a,DependentVariable):
+                    stack.extend(b for _,b in a.depends)
+
+        # set edges between all nodes
+        for a in g:
+            if isinstance(a,DependentVariable):
+                for p,b in a.depends:
+                    # add new edge for p if not already present, otherwise append to params list
+                    if (a,b) not in g.edges:
+                        g.add_edge(a,b,params=[p])
+                    else:
+                        g.edges[a,b]['params'].append(p)
```

Comment:
Now that unit tests are in place, I think that these two loops can probably be fused. I left them separate because adding edges to the graph automatically adds nodes, so this was making it messy to keep track of visited nodes. 

I would suggest adding a `visited` set, and using this instead:
```suggestion
        # discover all variables (nodes) by quasi-depth first search
        g = nx.DiGraph()
        stack = [self]
        visited = set()
        while stack:
            a = stack.pop()
            if a not in visited:
                g.add_node(a)
                visited.add(a)
                if isinstance(a,DependentVariable):
                    for p,b in a.depends:
                        stack.append(b)
                        # add new edge for p if not already present, otherwise append to params list
                        if (a,b) not in g.edges:
                            g.add_edge(a,b,params=[p])
                        else:
                            g.edges[a,b]['params'].append(p)
```

You can probably also remove the explicit `g.add_node(a)`, but this might be less readable.